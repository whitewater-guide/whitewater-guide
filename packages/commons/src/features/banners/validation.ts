import isNumber from 'lodash/isNumber';
import { customStruct } from '../../utils/validation';
import { BannerKind, BannerPlacement } from './types';

const struct = customStruct({
  ratio: (value: any) => {
    if (!isNumber(value)) {
      return 'Ratio must be a number';
    }
    if (value > 10 || value < 3) {
      return 'Ratio must be a number between 10 and 3';
    }
    return true;
  },
});

const BannerSourceStruct = struct.object({
  kind: struct.enum(Object.values(BannerKind)),
  ratio: 'ratio|null',
  src: 'nonEmptyString',
});

export const BannerInputStruct = struct.object({
  id: 'uuid|null',
  slug: 'slug',
  name: 'nonEmptyString',
  priority: 'integer',
  enabled: 'boolean',
  placement: struct.enum(Object.values(BannerPlacement)),
  source: BannerSourceStruct,
  link: 'https|null',
  extras: 'object|null',
  regions: ['node'],
  groups: ['node'],
});
