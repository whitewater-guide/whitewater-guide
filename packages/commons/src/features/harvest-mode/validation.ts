import { struct } from 'superstruct';
import { HarvestMode } from './types';

export const HarvestModeStruct = struct.enum([HarvestMode.ONE_BY_ONE, HarvestMode.ALL_AT_ONCE])
