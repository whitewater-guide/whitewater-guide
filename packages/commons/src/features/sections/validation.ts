// tslint:disable-next-line
import * as Joi from 'joi';
import { CoordinateSchema, PointInputSchema } from '../points';
import { Durations } from './types';

export const GaugeBindingSchema = Joi.object().keys({
  minimum: Joi.number().allow(null),
  maximum: Joi.number().allow(null),
  optimum: Joi.number().allow(null),
  impossible: Joi.number().allow(null),
  approximate: Joi.boolean().allow(null),
});

export const SectionInputSchema = Joi.object().keys({
  id: Joi.string().guid().allow(null),
  name: Joi.string().min(2).max(100),
  altNames: Joi.array().items(Joi.string().min(2)).allow(null),
  description: Joi.string().allow(null),
  season: Joi.string().allow(null),
  seasonNumeric: Joi.array().max(24).items(Joi.number().min(0).max(23)),

  river: Joi.object().keys({
    id: Joi.string().guid().allow(null),
  }),
  gauge: Joi.object().keys({ id: Joi.string().guid() }).allow(null),
  levels: GaugeBindingSchema.allow(null),
  flows: GaugeBindingSchema.allow(null),
  flowsText: Joi.string().allow(null),

  shape: Joi.array().items(CoordinateSchema).min(2),
  distance: Joi.number().positive().allow(null),
  drop: Joi.number().positive().allow(null),
  duration: Joi.any().valid(Array.from(Durations.keys())).allow(null),
  difficulty: Joi.number().min(0).max(5),
  difficultyXtra: Joi.string().max(32).allow(null),
  rating: Joi.number().min(0).max(5).allow(null),
  tags: Joi.array().items(Joi.object().keys({
    id: Joi.string(),
  })),
  pois: Joi.array().items(PointInputSchema),
});

// description is draft.js EditorState
export const SectionFormSchema = SectionInputSchema.keys({
  description: Joi.any(),
});
