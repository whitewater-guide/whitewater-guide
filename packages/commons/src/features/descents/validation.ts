import { yupTypes } from '@whitewater-guide/validation';
import * as yup from 'yup';
import { DescentLevelInput, DescentInput } from './types';

export const DescentLevelInputSchema = yup
  .object<DescentLevelInput>({
    unit: yup
      .string()
      .max(20)
      .notRequired()
      .nullable(),
    value: yup.number().required(),
  })
  .strict(true)
  .noUnknown();

export const DescentInputSchema = yup
  .object<DescentInput>({
    id: yupTypes
      .uuid()
      .notRequired()
      .nullable(),
    sectionId: yupTypes
      .uuid()
      .required()
      .nullable(false),
    startedAt: yup.date().required() as any,
    duration: yup
      .number()
      .integer()
      .notRequired()
      .nullable(),
    level: DescentLevelInputSchema.clone()
      .notRequired()
      .nullable(),
    comment: yup
      .string()
      .notRequired()
      .nullable(),
    public: yup
      .bool()
      .notRequired()
      .nullable(),
  })
  .strict(true)
  .noUnknown();
