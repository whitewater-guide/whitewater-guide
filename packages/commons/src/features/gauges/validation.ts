import { baseStruct } from '../../utils/validation';
import { PointInputStruct } from '../points';

const GaugeInputFields = {
  id: baseStruct.union(['uuid', 'null']),
  name: 'nonEmptyString',
  code: 'nonEmptyVarchar',
  levelUnit: 'nonEmptyVarchar|null',
  flowUnit: 'nonEmptyVarchar|null',
  location: baseStruct.union([PointInputStruct, 'null']),
  requestParams: 'object|null',
  cron: baseStruct.union(['cron', 'null']),
  url: baseStruct.union(['url', 'null', baseStruct.literal('')]),
  source: 'node',
};

export const GaugeInputStruct = baseStruct.object(GaugeInputFields);

export const GaugeFormStruct = baseStruct.object({
  ...GaugeInputFields,
  requestParams: baseStruct.union(['jsonString', 'null']),
});
