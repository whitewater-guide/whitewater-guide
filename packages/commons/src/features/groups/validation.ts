import * as yup from 'yup';

import { GroupInput } from './types';
import { yupTypes } from '@whitewater-guide/validation';

const GROUP_SKU = /^group\.\w{3,}$/;

export const GroupInputSchema = yup
  .object<GroupInput>({
    id: yupTypes.uuid(true),
    name: yupTypes.nonEmptyString(),
    sku: yup
      .string()
      .defined()
      .max(255)
      .matches(GROUP_SKU, 'yup:string.sku')
      .nullable(),
  })
  .strict(true)
  .noUnknown();
