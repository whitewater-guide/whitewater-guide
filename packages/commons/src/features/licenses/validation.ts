import { yupTypes } from '@whitewater-guide/validation';
import * as yup from 'yup';

import { License } from './types';

export const LicenseInputSchema: yup.SchemaOf<License> = yup
  .object({
    url: yup.string().url().nullable(true),
    name: yup.string().min(2).defined().nullable(false),
    slug: yupTypes.slug().nullable(true),
  })
  .strict(true)
  .noUnknown();
