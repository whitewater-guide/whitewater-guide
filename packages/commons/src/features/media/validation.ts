import * as yup from 'yup';

import { MediaInput, MediaKind } from './types';

import { yupTypes } from '@whitewater-guide/validation';

export const MediaInputSchema = yup
  .object<MediaInput>({
    id: yupTypes
      .uuid()
      .defined()
      .nullable(),
    description: yup
      .string()
      .defined()
      .nullable(),
    copyright: yup
      .string()
      .defined()
      .nullable(),
    url: yupTypes
      .nonEmptyString()
      .defined()
      .nullable(false),
    kind: yup
      .mixed()
      .defined()
      .oneOf(Object.values(MediaKind)),
    resolution: yup
      .array()
      .defined()
      .min(2)
      .max(2)
      .of(
        yup
          .number()
          .defined()
          .integer()
          .positive(),
      )
      .nullable(),
    weight: yup
      .number()
      .defined()
      .integer()
      .nullable(),
  })
  .strict(true)
  .noUnknown();
