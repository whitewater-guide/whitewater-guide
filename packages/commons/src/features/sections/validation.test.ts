import { NEW_ID } from '../../apollo';
import { createSafeValidator } from '../../validation';
import { MediaKind } from '../media';
import {
  Duration,
  GaugeBinding,
  SectionAdminSettings,
  SectionInput,
} from './types';
import {
  GaugeBindingSchema,
  SectionAdminSettingsSchema,
  SectionInputSchema,
} from './validation';

describe('GaugeBinding', () => {
  const validator = createSafeValidator(GaugeBindingSchema);
  type TestValue = [string, GaugeBinding];

  const correct: GaugeBinding = {
    minimum: -10,
    maximum: 10,
    optimum: 10,
    impossible: 10,
    approximate: true,
    formula: 'x + 10',
  };

  const correctValues: TestValue[] = [
    ['full value', correct],
    [
      'null value',
      {
        minimum: null,
        maximum: null,
        optimum: null,
        impossible: null,
        approximate: null,
        formula: null,
      },
    ],
    ['empty value', {}],
  ];

  const incorrectValues: TestValue[] = [
    ['incorrect formula', { ...correct, formula: 'x *' }],
    ['too many variables', { ...correct, formula: 'x * y + 2' }],
    ['bad variable', { ...correct, formula: 'X + 10' }],
    ['extra fields', { ...correct, foo: 'X + 10' } as any],
  ];

  it.each(correctValues)('should be valid for %s', (_, value) => {
    expect(validator(value)).toBeNull();
  });

  it.each(incorrectValues)('should be invalid for %s', (_, value) => {
    const error = validator(value);
    expect(error).not.toBeNull();
    expect(error).toMatchSnapshot();
  });
});

describe('SectionInput', () => {
  const validator = createSafeValidator(SectionInputSchema);
  type TestValue = [string, SectionInput];

  const correct: SectionInput = {
    id: 'eda6bd7e-c34a-11e8-a355-529269fb1459',
    name: 'Section',
    altNames: ['Alt name 1', 'Alt name 2'],
    description: 'Section description',
    season: 'When it rains',
    seasonNumeric: [10, 11, 12],
    river: { id: '1aa9fd9a-c34b-11e8-a355-529269fb1459' },
    gauge: { id: '2bedcd8e-c34b-11e8-a355-529269fb1459' },
    levels: { minimum: 10, optimum: null },
    flows: { maximum: 11, approximate: true },
    flowsText: 'unknown',
    shape: [
      [0, 0, 0],
      [0.1, 1, 0],
    ],
    distance: 10,
    drop: 100,
    duration: Duration.LAPS,
    difficulty: 3.5,
    difficultyXtra: 'X',
    rating: 2.5,
    tags: [{ id: '4x4' }, { id: 'portage' }],
    pois: [
      {
        id: 'b5678064-c34b-11e8-a355-529269fb1459',
        coordinates: [1, 1.3, 200],
        description: 'poi description',
        name: 'some poi',
        kind: 'portage',
      },
    ],
    media: [
      {
        id: '9f0ff4b6-c258-11e8-a355-529269fb1459',
        description: 'description',
        kind: MediaKind.photo,
        resolution: [800, 600],
        url: 'jhakjfh',
        weight: -3,
        copyright: 'Copyright',
      },
    ],
    hidden: false,
    helpNeeded: 'fix duration',
  };

  const correctValues: TestValue[] = [
    ['full value', correct],
    [
      'null value',
      {
        ...correct,
        id: null,
        altNames: null,
        description: null,
        difficultyXtra: null,
        distance: null,
        drop: null,
        duration: null,
        flows: null,
        flowsText: null,
        gauge: null,
        levels: null,
        rating: null,
        season: null,
        createdBy: null,
        suggestionId: null,
        helpNeeded: null,
        importId: null,
      },
    ],
    [
      'with createdBy',
      { ...correct, createdBy: '880e1a40-bd02-11e9-9cb5-2a2ae2dbcce4' },
    ],
    [
      'with suggestionId',
      { ...correct, createdBy: '8ef42282-bd02-11e9-9cb5-2a2ae2dbcce4' },
    ],
    ['with importId', { ...correct, importId: 'whatever' }],
    ['empty description', { ...correct, description: '' }],
    ['empty season', { ...correct, season: '' }],
    ['empty season numeric', { ...correct, seasonNumeric: [] }],
    ['empty flows text', { ...correct, flowsText: '' }],
    ['empty difficultyXtra', { ...correct, difficultyXtra: '' }],
    [
      'new river',
      {
        ...correct,
        river: { id: NEW_ID, name: 'New river' },
        region: { id: 'cf8c20ac-ae06-11e9-a2a3-2a2ae2dbcce4' },
      },
    ],
    ['empty media', { ...correct, media: [] }],
  ];

  const incorrectValues: TestValue[] = [
    ['bad id', { ...correct, id: 'foo' }],
    ['bad name', { ...correct, name: '' }],
    ['bad alt name', { ...correct, altNames: ['foo', '', 'bar'] }],
    [
      'bad season numeric (bad half-month)',
      { ...correct, seasonNumeric: [10, 100, -10, 11] },
    ],
    [
      'bad season numeric (too many items)',
      { ...correct, seasonNumeric: Array(30).fill(10) },
    ],
    ['bad river id', { ...correct, river: { id: 'foo' } }],
    ['bad gauge id', { ...correct, gauge: { id: 'foo' } }],
    ['bad shape (not enough points)', { ...correct, shape: [[1, 2, 3]] }],
    [
      'bad shape (bad point)',
      {
        ...correct,
        shape: [
          [1, 2, 3],
          [300, 1, 1],
          [2, 3, 4],
        ],
      },
    ],
    ['bad distance', { ...correct, distance: -100 }],
    ['bad drop', { ...correct, drop: -100 }],
    ['bad duration', { ...correct, duration: -100 }],
    ['bad difficulty (min)', { ...correct, difficulty: -0.5 }],
    ['bad difficulty (max)', { ...correct, difficulty: 7.5 }],
    ['bad difficulty (fraction)', { ...correct, difficulty: 4.4 }],
    ['bad extra difficulty', { ...correct, difficultyXtra: 'x'.repeat(50) }],
    ['bad rating (min)', { ...correct, rating: -0.5 }],
    ['bad rating (max)', { ...correct, rating: 5.5 }],
    ['bad rating (fraction)', { ...correct, rating: 4.4 }],
    ['bad tags', { ...correct, tags: ['foo', '', 'bar'] }],
    [
      'bad pois',
      {
        ...correct,
        pois: [
          {
            id: null,
            description: null,
            kind: 'dam',
            name: null,
            coordinates: [1, 1, 1],
          },
          {
            id: 'foo',
            description: null,
            kind: 'dam',
            name: null,
            coordinates: [1, 1, 1],
          },
          {
            id: null,
            description: null,
            kind: 'dam',
            name: null,
            coordinates: [1, 2, 1],
          },
        ],
      },
    ],
    ['null media', { ...correct, media: null as any }],
    [
      'incorrect media',
      { ...correct, media: [{ ...correct.media![0], id: 'foo' }] },
    ],
    ['extra fields', { ...correct, foo: 'bar' } as any],
    ['new river without name', { ...correct, river: { id: NEW_ID } }],
  ];

  it.each(correctValues)('should be valid for %s', (_, value) => {
    expect(validator(value)).toBeNull();
  });

  it.each(incorrectValues)('should be invalid for %s', (_, value) => {
    const error = validator(value);
    expect(error).not.toBeNull();
    expect(error).toMatchSnapshot();
  });
});

describe('SectionAdminSettings', () => {
  const validator = createSafeValidator(SectionAdminSettingsSchema);
  type TestValue = [string, SectionAdminSettings];

  const correct: SectionAdminSettings = { demo: true };

  const correctValues: TestValue[] = [['full value', correct]];
  const incorrectValues: TestValue[] = [
    ['demo undefined', {} as any],
    ['extra fields', { ...correct, foo: 'bar' } as any],
  ];

  it.each(correctValues)('should be valid for %s', (_, value) => {
    expect(validator(value)).toBeNull();
  });

  it.each(incorrectValues)('should be invalid for %s', (_, value) => {
    const error = validator(value);
    expect(error).not.toBeNull();
    expect(error).toMatchSnapshot();
  });
});
