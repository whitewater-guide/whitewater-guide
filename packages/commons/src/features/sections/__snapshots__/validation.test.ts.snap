// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SectionAdminSettings should be invalid for extra fields 1`] = `
Object {
  "foo": "Expected a value of type \`undefined\` for \`foo\` but received \`\\"bar\\"\`.",
}
`;

exports[`SectionFormInput should be invalid for bad alt name 1`] = `
Object {
  "altNames": Array [
    undefined,
    "Value must be non-empty string",
  ],
}
`;

exports[`SectionFormInput should be invalid for bad difficulty (fraction) 1`] = `
Object {
  "difficulty": "Expected a value of type \`0 | 0.5 | 1 | 1.5 | 2 | 2.5 | 3 | 3.5 | 4 | 4.5 | 5\` for \`difficulty\` but received \`4.4\`.",
}
`;

exports[`SectionFormInput should be invalid for bad difficulty (max) 1`] = `
Object {
  "difficulty": "Expected a value of type \`0 | 0.5 | 1 | 1.5 | 2 | 2.5 | 3 | 3.5 | 4 | 4.5 | 5\` for \`difficulty\` but received \`5.5\`.",
}
`;

exports[`SectionFormInput should be invalid for bad difficulty (min) 1`] = `
Object {
  "difficulty": "Expected a value of type \`0 | 0.5 | 1 | 1.5 | 2 | 2.5 | 3 | 3.5 | 4 | 4.5 | 5\` for \`difficulty\` but received \`-0.5\`.",
}
`;

exports[`SectionFormInput should be invalid for bad distance 1`] = `
Object {
  "distance": "Value must be positive number",
}
`;

exports[`SectionFormInput should be invalid for bad drop 1`] = `
Object {
  "drop": "Value must be positive number",
}
`;

exports[`SectionFormInput should be invalid for bad duration 1`] = `
Object {
  "duration": "Expected a value of type \`null | 10 | 20 | 30 | 40 | 50\` for \`duration\` but received \`-100\`.",
}
`;

exports[`SectionFormInput should be invalid for bad extra difficulty 1`] = `
Object {
  "difficultyXtra": "Must be string no longer than 32 chars",
}
`;

exports[`SectionFormInput should be invalid for bad gauge id 1`] = `
Object {
  "gauge": "Value must be node with id",
}
`;

exports[`SectionFormInput should be invalid for bad id 1`] = `
Object {
  "id": "Incorrect UUID",
}
`;

exports[`SectionFormInput should be invalid for bad name 1`] = `
Object {
  "name": "Value must be non-empty string",
}
`;

exports[`SectionFormInput should be invalid for bad pois 1`] = `
Object {
  "pois": Array [
    undefined,
    Object {
      "id": "Incorrect UUID",
    },
  ],
}
`;

exports[`SectionFormInput should be invalid for bad rating (fraction) 1`] = `
Object {
  "rating": "Expected a value of type \`null | 0 | 0.5 | 1 | 1.5 | 2 | 2.5 | 3 | 3.5 | 4 | 4.5 | 5\` for \`rating\` but received \`4.4\`.",
}
`;

exports[`SectionFormInput should be invalid for bad rating (max) 1`] = `
Object {
  "rating": "Expected a value of type \`null | 0 | 0.5 | 1 | 1.5 | 2 | 2.5 | 3 | 3.5 | 4 | 4.5 | 5\` for \`rating\` but received \`5.5\`.",
}
`;

exports[`SectionFormInput should be invalid for bad rating (min) 1`] = `
Object {
  "rating": "Expected a value of type \`null | 0 | 0.5 | 1 | 1.5 | 2 | 2.5 | 3 | 3.5 | 4 | 4.5 | 5\` for \`rating\` but received \`-0.5\`.",
}
`;

exports[`SectionFormInput should be invalid for bad river id 1`] = `
Object {
  "river": "Value must be node with id",
}
`;

exports[`SectionFormInput should be invalid for bad season numeric (bad half-month) 1`] = `
Object {
  "seasonNumeric": Array [
    undefined,
    "Must be integer between 0 and 23",
    "Must be integer between 0 and 23",
  ],
}
`;

exports[`SectionFormInput should be invalid for bad season numeric (too many items) 1`] = `
Object {
  "seasonNumeric": "Maximal length is 24",
}
`;

exports[`SectionFormInput should be invalid for bad shape (bad point) 1`] = `
Object {
  "shape": Array [
    undefined,
    Array [
      "Longitude value must be between -180 and 180",
    ],
  ],
}
`;

exports[`SectionFormInput should be invalid for bad shape (not enough points) 1`] = `
Object {
  "shape": "Minimal length is 2",
}
`;

exports[`SectionFormInput should be invalid for bad tag 1`] = `
Object {
  "kayakingTags": Array [
    undefined,
    Object {
      "id": "Value must be valid slug",
    },
  ],
}
`;

exports[`SectionFormInput should be invalid for extra fields 1`] = `
Object {
  "foo": "Expected a value of type \`undefined\` for \`foo\` but received \`\\"bar\\"\`.",
}
`;

exports[`SectionFormInput should be invalid for present tag field 1`] = `
Object {
  "tags": "Expected a value of type \`undefined\` for \`tags\` but received \`[{\\"id\\":\\"4x4\\"}]\`.",
}
`;

exports[`SectionInput should be invalid for bad alt name 1`] = `
Object {
  "altNames": Array [
    undefined,
    "Value must be non-empty string",
  ],
}
`;

exports[`SectionInput should be invalid for bad difficulty (fraction) 1`] = `
Object {
  "difficulty": "Expected a value of type \`0 | 0.5 | 1 | 1.5 | 2 | 2.5 | 3 | 3.5 | 4 | 4.5 | 5\` for \`difficulty\` but received \`4.4\`.",
}
`;

exports[`SectionInput should be invalid for bad difficulty (max) 1`] = `
Object {
  "difficulty": "Expected a value of type \`0 | 0.5 | 1 | 1.5 | 2 | 2.5 | 3 | 3.5 | 4 | 4.5 | 5\` for \`difficulty\` but received \`5.5\`.",
}
`;

exports[`SectionInput should be invalid for bad difficulty (min) 1`] = `
Object {
  "difficulty": "Expected a value of type \`0 | 0.5 | 1 | 1.5 | 2 | 2.5 | 3 | 3.5 | 4 | 4.5 | 5\` for \`difficulty\` but received \`-0.5\`.",
}
`;

exports[`SectionInput should be invalid for bad distance 1`] = `
Object {
  "distance": "Value must be positive number",
}
`;

exports[`SectionInput should be invalid for bad drop 1`] = `
Object {
  "drop": "Value must be positive number",
}
`;

exports[`SectionInput should be invalid for bad duration 1`] = `
Object {
  "duration": "Expected a value of type \`null | 10 | 20 | 30 | 40 | 50\` for \`duration\` but received \`-100\`.",
}
`;

exports[`SectionInput should be invalid for bad extra difficulty 1`] = `
Object {
  "difficultyXtra": "Must be string no longer than 32 chars",
}
`;

exports[`SectionInput should be invalid for bad gauge id 1`] = `
Object {
  "gauge": "Value must be node with id",
}
`;

exports[`SectionInput should be invalid for bad id 1`] = `
Object {
  "id": "Incorrect UUID",
}
`;

exports[`SectionInput should be invalid for bad name 1`] = `
Object {
  "name": "Value must be non-empty string",
}
`;

exports[`SectionInput should be invalid for bad pois 1`] = `
Object {
  "pois": Array [
    undefined,
    Object {
      "id": "Incorrect UUID",
    },
  ],
}
`;

exports[`SectionInput should be invalid for bad rating (fraction) 1`] = `
Object {
  "rating": "Expected a value of type \`null | 0 | 0.5 | 1 | 1.5 | 2 | 2.5 | 3 | 3.5 | 4 | 4.5 | 5\` for \`rating\` but received \`4.4\`.",
}
`;

exports[`SectionInput should be invalid for bad rating (max) 1`] = `
Object {
  "rating": "Expected a value of type \`null | 0 | 0.5 | 1 | 1.5 | 2 | 2.5 | 3 | 3.5 | 4 | 4.5 | 5\` for \`rating\` but received \`5.5\`.",
}
`;

exports[`SectionInput should be invalid for bad rating (min) 1`] = `
Object {
  "rating": "Expected a value of type \`null | 0 | 0.5 | 1 | 1.5 | 2 | 2.5 | 3 | 3.5 | 4 | 4.5 | 5\` for \`rating\` but received \`-0.5\`.",
}
`;

exports[`SectionInput should be invalid for bad river id 1`] = `
Object {
  "river": "Value must be node with id",
}
`;

exports[`SectionInput should be invalid for bad season numeric (bad half-month) 1`] = `
Object {
  "seasonNumeric": Array [
    undefined,
    "Must be integer between 0 and 23",
    "Must be integer between 0 and 23",
  ],
}
`;

exports[`SectionInput should be invalid for bad season numeric (too many items) 1`] = `
Object {
  "seasonNumeric": "Maximal length is 24",
}
`;

exports[`SectionInput should be invalid for bad shape (bad point) 1`] = `
Object {
  "shape": Array [
    undefined,
    Array [
      "Longitude value must be between -180 and 180",
    ],
  ],
}
`;

exports[`SectionInput should be invalid for bad shape (not enough points) 1`] = `
Object {
  "shape": "Minimal length is 2",
}
`;

exports[`SectionInput should be invalid for bad tags 1`] = `
Object {
  "tags": Array [
    "Expected a value of type \`{id}\` for \`tags.0\` but received \`\\"foo\\"\`.",
    "Incorrect value",
    "Incorrect value",
  ],
}
`;

exports[`SectionInput should be invalid for extra fields 1`] = `
Object {
  "foo": "Expected a value of type \`undefined\` for \`foo\` but received \`\\"bar\\"\`.",
}
`;
